src/command.rs

```diff
@@ -0,0 +1,304 @@
+        .push(Dependency::new(registry, repository, package, version));
+
+    manifest.write().await
+}
+
+/// Removes a dependency from this project
+pub async fn remove(package: PackageName) -> eyre::Result<()> {
+    let mut manifest = Manifest::read().await?;
+
+    let dependency = manifest
+        .dependencies
+        .iter()
+        .find(|d| d.package == package)
+        .wrap_err(eyre::eyre!(
+            "Unable to remove unknown dependency {package:?}"
+        ))?
+        .to_owned();
+
+    manifest.dependencies.retain(|d| *d != dependency);
+
+    PackageStore::uninstall(&dependency.package).await.ok();
+
+    manifest.write().await
+}
+
+/// Packages the api and writes it to the filesystem
+pub async fn package(directory: String, dry_run: bool) -> eyre::Result<()> {
+    let package = PackageStore::release()
+        .await
+        .wrap_err("Failed to create release")?;
+
+    let path = Path::new(&directory).join(format!(
+        "{}-{}.tgz",
+        package.manifest.package.name, package.manifest.package.version
+    ));
+
+    if !dry_run {
+        std::fs::write(path, package.tgz).wrap_err("failed to write package to filesystem")?;
+    }
+
+    Ok(())
+}
+
+/// Publishes the api package to the registry
+pub async fn publish(
+    credentials: Credentials,
+    registry: RegistryUri,
+    repository: String,
+    allow_dirty: bool,
+    dry_run: bool,
+) -> eyre::Result<()> {
+    if let Ok(repository) = git2::Repository::discover(Path::new(".")) {
+        let statuses = repository
+            .statuses(None)
+            .wrap_err("Failed to get git status")?;
+
+        if !allow_dirty && !statuses.is_empty() {
+            tracing::error!("{} files in the working directory contain changes that were not yet committed into git:\n", statuses.len());
+
+            statuses
+                .iter()
+                .for_each(|s| tracing::error!("{}", s.path().unwrap_or_default()));
+
+            tracing::error!("\nTo proceed with publishing despite the uncommitted changes, pass the `--allow-dirty` flag\n");
+
+            eyre::bail!("Unable to publish a dirty git repository");
+        }
+    }
+
+    let artifactory = Artifactory::new(Arc::new(credentials), registry);
+
+    let package = PackageStore::release()
+        .await
+        .wrap_err("Failed to create release")?;
+
+    if dry_run {
+        tracing::warn!(":: aborting upload due to dry run");
+        return Ok(());
+    }
+
+    artifactory.publish(package, repository).await?;
+
+    Ok(())
+}
+
+/// Installs dependencies
+pub async fn install(credentials: Credentials) -> eyre::Result<()> {
+    let credentials = Arc::new(credentials);
+
+    let manifest = Manifest::read().await?;
+
+    let lockfile = Lockfile::read_or_default().await?;
+
+    let dependency_tree =
+        DependencyGraph::from_manifest(&manifest, &lockfile, &credentials).await?;
+
+    let mut locked = Vec::new();
+
+    #[async_recursion]
+    async fn process_recursively(
```

Comment:
agree, feels somewhat hard to track

Reply:
This isn't a resolver, it's just traversing the dependency graph that was build previously. I can improve the naming but I don't think it's a good idea to unify the two or we'd end up with a situation where we fail validation but have already installed some packages.

Reply:
I also considered implementing a `DependencyGraph::traverse()` method and providing it a closure here but couldn't figure out a clean way to keep track of the prefix with that approach.

Reply:
TBH I kinda want to reimplement this with an iterative approach to avoid the dependency on `async_recursion` but I figured this would be easier to reason about for a first implementation.